File: src/utils/confirmationDeadlines.ts
import { addDays, subDays, subHours, isAfter, isBefore, formatDistanceToNow, addHours } from 'date-fns';
import { AutoReleaseSettings, DEFAULT_AUTO_RELEASE_SETTINGS } from '@/types/postponement';

export interface ConfirmationDeadlines {
  confirmation_deadline: string;
  autorelease_at: string;
  confirmation_status: 'pending' | 'confirmed' | 'auto_released';
}

export interface DateRangeDeadlines extends ConfirmationDeadlines {
  visitDateFrom: string;
  visitDateTo: string;
  effectiveVisitDate: string;
}

/**
 * Get auto-release settings from localStorage or use defaults
 */
export function getAutoReleaseSettings(): AutoReleaseSettings {
  try {
    const stored = localStorage.getItem('autoReleaseSettings');
    if (stored) {
      return { ...DEFAULT_AUTO_RELEASE_SETTINGS, ...JSON.parse(stored) };
    }
  } catch (e) {
    console.error('Failed to parse auto-release settings:', e);
  }
  return DEFAULT_AUTO_RELEASE_SETTINGS;
}

/**
 * Save auto-release settings to localStorage
 */
export function saveAutoReleaseSettings(settings: Partial<AutoReleaseSettings>): void {
  const current = getAutoReleaseSettings();
  const updated = { ...current, ...settings };
  localStorage.setItem('autoReleaseSettings', JSON.stringify(updated));
}

/**
 * Calculate confirmation deadlines based on scheduled visit date
 * Uses configurable timing from settings
 */
export function calculateConfirmationDeadlines(
  visitDate: string | Date, 
  settings?: AutoReleaseSettings
): ConfirmationDeadlines {
  const config = settings || getAutoReleaseSettings();
  const visit = new Date(visitDate);
  
  // Confirmation deadline is configurable hours before the visit
  const confirmationDeadline = subHours(visit, config.confirmationHoursBeforeVisit);
  
  // Auto-release triggers configurable hours before the visit if not confirmed
  const autoreleaseAt = subHours(visit, config.releaseHoursBeforeVisit);
  
  return {
    confirmation_deadline: confirmationDeadline.toISOString(),
    autorelease_at: autoreleaseAt.toISOString(),
    confirmation_status: 'pending',
  };
}

/**
 * Calculate deadlines for date range visits (DM/GFA activities)
 * Uses the start date (dateFrom) as the effective date for deadline calculations
 */
export function calculateDateRangeDeadlines(
  dateFrom: string | Date,
  dateTo: string | Date,
  settings?: AutoReleaseSettings
): DateRangeDeadlines {
  const config = settings || getAutoReleaseSettings();
  const visitStart = new Date(dateFrom);
  const visitEnd = new Date(dateTo);
  
  // For date range visits, use the start date for deadline calculations
  const baseDeadlines = calculateConfirmationDeadlines(dateFrom, config);
  
  return {
    ...baseDeadlines,
    visitDateFrom: visitStart.toISOString(),
    visitDateTo: visitEnd.toISOString(),
    effectiveVisitDate: visitStart.toISOString(),
  };
}

/**
 * Get reminder times for a confirmation deadline
 * Returns timestamps for: 2 days before, 1 day before, 12 hours before
 */
export function getReminderTimes(confirmationDeadline: string | Date): {
  twoDaysBefore: Date;
  oneDayBefore: Date;
  twelveHoursBefore: Date;
} {
  const deadline = new Date(confirmationDeadline);
  
  return {
    twoDaysBefore: subDays(deadline, 2),
    oneDayBefore: subDays(deadline, 1),
    twelveHoursBefore: subHours(deadline, 12),
  };
}

/**
 * Check if confirmation deadline has passed
 */
export function isDeadlinePassed(confirmationDeadline: string | Date): boolean {
  return isAfter(new Date(), new Date(confirmationDeadline));
}

/**
 * Check if auto-release should trigger
 */
export function shouldAutoRelease(
  autoreleaseAt: string | Date,
  confirmationStatus: string
): boolean {
  if (confirmationStatus === 'confirmed') return false;
  return isAfter(new Date(), new Date(autoreleaseAt));
}

/**
 * Get human-readable time remaining until deadline
 */
export function getTimeRemaining(deadline: string | Date): string {
  const deadlineDate = new Date(deadline);
  if (isAfter(new Date(), deadlineDate)) {
    return 'Expired';
  }
  return formatDistanceToNow(deadlineDate, { addSuffix: true });
}

/**
 * Get urgency level based on time remaining
 */
export function getDeadlineUrgency(deadline: string | Date): 'normal' | 'warning' | 'critical' | 'expired' {
  const deadlineDate = new Date(deadline);
  const now = new Date();
  
  if (isAfter(now, deadlineDate)) {
    return 'expired';
  }
  
  const hoursRemaining = (deadlineDate.getTime() - now.getTime()) / (1000 * 60 * 60);
  
  if (hoursRemaining <= 12) {
    return 'critical';
  }
  if (hoursRemaining <= 24) {
    return 'warning';
  }
  return 'normal';
}

File: mobile_flutter/PACT_mobile-PACT-MAIN-shorebird/lib/utils/confirmation_deadlines.dart
import 'package:shared_preferences/shared_preferences.dart';
import 'dart:convert';

class AutoReleaseSettings {
  final int confirmationHoursBeforeVisit;
  final int releaseHoursBeforeVisit;
  final bool autoReleaseEnabled;

  const AutoReleaseSettings({
    this.confirmationHoursBeforeVisit = 24,
    this.releaseHoursBeforeVisit = 12,
    this.autoReleaseEnabled = true,
  });

  factory AutoReleaseSettings.fromJson(Map<String, dynamic> json) {
    return AutoReleaseSettings(
      confirmationHoursBeforeVisit:
          json['confirmationHoursBeforeVisit'] as int? ?? 24,
      releaseHoursBeforeVisit: json['releaseHoursBeforeVisit'] as int? ?? 12,
      autoReleaseEnabled: json['autoReleaseEnabled'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() => {
    'confirmationHoursBeforeVisit': confirmationHoursBeforeVisit,
    'releaseHoursBeforeVisit': releaseHoursBeforeVisit,
    'autoReleaseEnabled': autoReleaseEnabled,
  };
}

class ConfirmationDeadlines {
  final String confirmationDeadline;
  final String autoreleaseAt;
  final String confirmationStatus;

  ConfirmationDeadlines({
    required this.confirmationDeadline,
    required this.autoreleaseAt,
    this.confirmationStatus = 'pending',
  });

  Map<String, dynamic> toJson() => {
    'confirmation_deadline': confirmationDeadline,
    'autorelease_at': autoreleaseAt,
    'confirmation_status': confirmationStatus,
  };
}

class DateRangeDeadlines extends ConfirmationDeadlines {
  final String visitDateFrom;
  final String visitDateTo;
  final String effectiveVisitDate;

  DateRangeDeadlines({
    required super.confirmationDeadline,
    required super.autoreleaseAt,
    super.confirmationStatus,
    required this.visitDateFrom,
    required this.visitDateTo,
    required this.effectiveVisitDate,
  });

  @override
  Map<String, dynamic> toJson() => {
    ...super.toJson(),
    'visitDateFrom': visitDateFrom,
    'visitDateTo': visitDateTo,
    'effectiveVisitDate': effectiveVisitDate,
  };
}

class ReminderTimes {
  final DateTime twoDaysBefore;
  final DateTime oneDayBefore;
  final DateTime twelveHoursBefore;

  ReminderTimes({
    required this.twoDaysBefore,
    required this.oneDayBefore,
    required this.twelveHoursBefore,
  });
}

class ConfirmationDeadlineUtils {
  static const String _settingsKey = 'autoReleaseSettings';
  static const AutoReleaseSettings defaultSettings = AutoReleaseSettings();

  static Future<AutoReleaseSettings> getAutoReleaseSettings() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final stored = prefs.getString(_settingsKey);
      if (stored != null) {
        final json = jsonDecode(stored) as Map<String, dynamic>;
        return AutoReleaseSettings.fromJson(json);
      }
    } catch (e) {
      print('Failed to parse auto-release settings: $e');
    }
    return defaultSettings;
  }

  static Future<void> saveAutoReleaseSettings(
    AutoReleaseSettings settings,
  ) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(_settingsKey, jsonEncode(settings.toJson()));
    } catch (e) {
      print('Failed to save auto-release settings: $e');
    }
  }

  static ConfirmationDeadlines calculateConfirmationDeadlines(
    DateTime visitDate, {
    AutoReleaseSettings? settings,
  }) {
    final config = settings ?? defaultSettings;

    final confirmationDeadline = visitDate.subtract(
      Duration(hours: config.confirmationHoursBeforeVisit),
    );

    final autoreleaseAt = visitDate.subtract(
      Duration(hours: config.releaseHoursBeforeVisit),
    );

    return ConfirmationDeadlines(
      confirmationDeadline: confirmationDeadline.toIso8601String(),
      autoreleaseAt: autoreleaseAt.toIso8601String(),
      confirmationStatus: 'pending',
    );
  }

  static Future<ConfirmationDeadlines> calculateConfirmationDeadlinesAsync(
    DateTime visitDate,
  ) async {
    final settings = await getAutoReleaseSettings();
    return calculateConfirmationDeadlines(visitDate, settings: settings);
  }

  static DateRangeDeadlines calculateDateRangeDeadlines(
    DateTime dateFrom,
    DateTime dateTo, {
    AutoReleaseSettings? settings,
  }) {
    final config = settings ?? defaultSettings;
    final baseDeadlines = calculateConfirmationDeadlines(
      dateFrom,
      settings: config,
    );

    return DateRangeDeadlines(
      confirmationDeadline: baseDeadlines.confirmationDeadline,
      autoreleaseAt: baseDeadlines.autoreleaseAt,
      confirmationStatus: baseDeadlines.confirmationStatus,
      visitDateFrom: dateFrom.toIso8601String(),
      visitDateTo: dateTo.toIso8601String(),
      effectiveVisitDate: dateFrom.toIso8601String(),
    );
  }

  static ReminderTimes getReminderTimes(DateTime confirmationDeadline) {
    return ReminderTimes(
      twoDaysBefore: confirmationDeadline.subtract(const Duration(days: 2)),
      oneDayBefore: confirmationDeadline.subtract(const Duration(days: 1)),
      twelveHoursBefore: confirmationDeadline.subtract(
        const Duration(hours: 12),
      ),
    );
  }

  static bool isConfirmationOverdue(DateTime confirmationDeadline) {
    return DateTime.now().isAfter(confirmationDeadline);
  }

  static bool shouldAutoRelease(DateTime autoreleaseAt) {
    return DateTime.now().isAfter(autoreleaseAt);
  }

  static String getConfirmationStatusLabel(String status) {
    switch (status) {
      case 'confirmed':
        return 'Confirmed';
      case 'auto_released':
        return 'Auto-Released';
      case 'pending':
      default:
        return 'Pending Confirmation';
    }
  }

  static Duration getTimeUntilDeadline(DateTime deadline) {
    return deadline.difference(DateTime.now());
  }

  static String formatTimeRemaining(Duration duration) {
    if (duration.isNegative) {
      return 'Overdue';
    }

    if (duration.inDays > 0) {
      return '${duration.inDays}d ${duration.inHours % 24}h remaining';
    } else if (duration.inHours > 0) {
      return '${duration.inHours}h ${duration.inMinutes % 60}m remaining';
    } else {
      return '${duration.inMinutes}m remaining';
    }
  }
}

const AutoReleaseSettings defaultAutoReleaseSettings = AutoReleaseSettings();
